parser/src/config/csv.rs

```diff
@@ -0,0 +1,158 @@
+        schema.into()
+    }
+}
+
+impl Into<u8> for Char {
+    fn into(self) -> u8 {
+        self.0
+    }
+}
+
+impl<'a> TryFrom<&'a str> for Char {
+    type Error = &'static str;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        if value.len() != 1 {
+            Err("expected a single ascii character")
+        } else {
+            Ok(Char(value.as_bytes()[0]))
+        }
+    }
+}
+
+impl<'de> de::Deserialize<'de> for Char {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = std::borrow::Cow::<'de, str>::deserialize(deserializer)?;
+        Char::try_from(s.as_ref()).map_err(|e| de::Error::custom(e))
+    }
+}
+
+impl ser::Serialize for Char {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let bytes = &[self.0];
+        let s = std::str::from_utf8(bytes)
+            .map_err(|_| ser::Error::custom("value is not valid utf-8"))?;
+        serializer.serialize_str(s)
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub enum LineEnding {
+    CRLF,
+    Other(Char),
+}
+
+impl JsonSchema for LineEnding {
+    fn schema_name() -> String {
+        "LineEnding".to_string()
+    }
+
+    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
+        let char_schema = gen.subschema_for::<Char>();
+        serde_json::from_value(serde_json::json!({
+            "description": "the character(s) that separates lines, which must either be a single character or '\r\n'",
+            "oneOf": [
+                { "enum": ["\r\n"] },
+                char_schema,
+            ]
+        })).unwrap()
+    }
+}
+
+impl<'a> TryFrom<&'a str> for LineEnding {
+    type Error = &'static str;
+
+    fn try_from(value: &'a str) -> Result<Self, Self::Error> {
+        if value == "\r\n" {
+            Ok(LineEnding::CRLF)
+        } else if value.len() == 1 {
+            Char::try_from(value).map(LineEnding::Other)
+        } else {
+            Err(r#"expected either a single character or the string "\r\n""#)
+        }
+    }
+}
+
+impl<'de> de::Deserialize<'de> for LineEnding {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let s = std::borrow::Cow::<'de, str>::deserialize(deserializer)?;
+        LineEnding::try_from(s.as_ref()).map_err(|e| de::Error::custom(e))
+    }
+}
+
+impl ser::Serialize for LineEnding {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match self {
+            LineEnding::CRLF => serializer.serialize_str("\r\n"),
+            LineEnding::Other(c) => c.serialize(serializer),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Default, JsonSchema, Serialize, Deserialize)]
+pub struct CharacterSeparatedConfig {
```

Comment:
The intent here is that none of this config should be required for any common cases. The parser should be able to handle most files without any configuration, and the happy-path tests intentionally don't provide any so that we can ensure this is the case. The options here are only required for files that are in some weird format, or where for some reason our best-effort guess proves incorrect. There are a number of reasons why that might happen, and I'll try to elaborate on specifics in the other comment threads, but the high level is just that there's no guarantee that we can guess correctly 100% of the time on these things.